File: algorithm-design-and-analysis/cs161-stanford/programming-section/week4/codes/quicksort.cpp
#include <iostream>
#include <vector>
#include <utility>
#include <cstdlib>

using namespace std;

typedef vector<int> VI;

VI arr;

int partition(int start, int end) {
  int pivot = start;
  bool flag = false;
  for (int i = start + 1; i < end; i++) {
    if (arr[i] < arr[start] || (arr[i] == arr[start] && (flag = !flag))) swap(arr[++pivot], arr[i]);
  }
  swap(arr[start], arr[pivot]);
  return pivot;
}

void quicksort(int start, int end) {
  if (start >= end - 1) return;
  swap(arr[start], arr[start + (rand() % (end - start))]);
  int pivot = partition(start, end);
  quicksort(start, pivot);
  quicksort(pivot + 1, end);
}

int main() {
  int n;
  cin >> n;
  arr = VI(n);
  for (int i = 0; i < n; i++) cin >> arr[i];
  
  quicksort(0, n);
  
  for (int i = 0; i < n; i++) cout << arr[i] << endl;
  
  return 0;
}

File: algorithm-design-and-analysis/cs161-stanford/programming-section/week4/codes/Quicksort.java
import java.util.*;


public class Quicksort {
  static int[] arr;
  static Random rnd = new Random();
  
  static void swap(int a, int b) {
    int temp = arr[a]; arr[a] = arr[b]; arr[b] = temp;
  }
  
  static int partition(int start, int end) {
    int pivot = start;
    boolean flag = false;
    for (int i = start + 1; i < end; i++) {
      if (arr[i] < arr[start] || (arr[i] == arr[start] && (flag = !flag))) swap(++pivot, i);
    }
    swap(start, pivot);
    return pivot;
  }
  
  static void quicksort(int start, int end) {
    if (start >= end - 1) return;
    swap(start, start + rnd.nextInt(end - start));
    int pivot = partition(start, end);
    quicksort(start, pivot);
    quicksort(pivot + 1, end);
  }
  
  public static void main(String[] args) {
    Scanner s = new Scanner(System.in);
    int n = s.nextInt();
    arr = new int[n];
    for (int i = 0; i < n; i++) arr[i] = s.nextInt();
    
    quicksort(0, n);
    
    for (int i = 0; i < n; i++) System.out.println(arr[i]);
  }
}

